### PR TITLE
Wide-code handling: horizontal pan, navigator collapse, pane zoom

### PR DESCRIPTION
**Pull request by JonasBaeumer**
## What

Long code lines were clipped by the split view with no way to see the rest (user feedback from live review). Wrapping was rejected — it destroys code readability — so this PR ships three composable mechanisms:

- **Horizontal pan**: `←`/`→` (or `H`/`L`, or trackpad horizontal scroll) shift the code 8 columns per step while the line-number gutter and `+`/`-` origin marker stay pinned. Clipped edges are marked with dim `‹` / `…` indicators; `0` resets. Pan state resets on file switch; inline comment rows keep wrapping instead of panning.
- **Navigator collapse**: `b` hides the file list, giving the diff the full pane width (annot/reviewr precedent).
- **Pane zoom**: `z` toggles herdr's own pane zoom via the CLI for distraction-free full-screen reading.

## How was it tested

- [x] `cargo test` — 39 tests green (new: pan/clip edge cases incl. pinned spans, both clip markers, overshoot; collapsed-navigator layout)
- [x] Build warning-free
- [ ] Live in herdr — reviewer: try `→` on a long line, `b`, and `z` in a real review

## Checklist

- [x] Build is warning-free
- [x] README update follows in the release PR (keybindings table)
- [x] No code copied from AGPL sources

## Dependency note

Adds `unicode-width` 0.2 and `unicode-segmentation` 1 (direct dependencies, pure Rust, no transitive additions — it was already in the tree via ratatui). Required because pan/clip must measure **terminal display columns** for arbitrary slices and per-char boundary decisions: ratatui's width facilities apply at render time to whole spans and are not exposed for the incremental per-scalar arithmetic the pan boundary logic needs. Using the same crate ratatui itself uses keeps our arithmetic identical to the render layer's.

### CODE REVIEW

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 1]
```diff
@@ -1235,13 +1283,24 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+    let inner_width = (area.width.saturating_sub(2)).max(1) as usize;
+
+    // Horizontal pan + clip indicators for code rows. Gutter and origin
+    // marker (the first two spans of a line row) stay pinned; hunk headers
+    // and placeholders (single-span rows) pan whole. Runs after the style
+    // patches (they only touch styles/first gutter char) and before comment
+    // weaving (comments wrap instead of panning).
+    for line in &mut lines {
+        let pinned = if line.spans.len() >= 3 { 2 } else { 0 };
+        pan_and_clip(line, app.diff.hscroll, inner_width, pinned);
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:1430]: Fixed in 16f0929: blank changed lines now always emit an (empty) content span, so the 3+-spans pin rule holds for them. Regression test `blank_changed_lines_keep_their_gutter_under_panning` pans an empty added line 16 columns and asserts gutter+marker survive — it fails without the fix.

File: src/ui.rs [hunk 2]
```diff
@@ -1271,6 +1330,61 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+                    chars[0] = '\u{2039}'; // ‹
+                    span.content = chars.into_iter().collect::<String>().into();
+                    break;
+                }
+            }
+        }
+    }
+
+    let total: usize = line.spans.iter().map(|s| s.content.chars().count()).sum();
+    if total > width {
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 01a0f8f: pan_and_clip now does all arithmetic in terminal display columns via unicode-width (pure Rust, keeping the no-C-deps rule). A wide char straddling the pan boundary drops whole with a pad space so columns stay aligned, and the ‹ marker preserves the replaced cell's width. Tests cover CJK pan, straddle, and right-clip (`pan_and_clip_counts_display_columns_not_chars`).

File: src/ui.rs [hunk 3]
```diff
@@ -972,6 +1005,13 @@ impl<'a> App<'a> {
                     }
                     KeyCode::Char('g') => self.diff.top(),
                     KeyCode::Char('G') => self.diff.bottom(row_count),
+                    KeyCode::Right | KeyCode::Char('L') => {
+                        self.diff.hscroll = (self.diff.hscroll + HSCROLL_STEP).min(1000)
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 3dbac37: the literal 1000 is gone — the pan cap is now derived per file at cache-build time (widest row's pannable display columns minus one), so any line's tail is reachable and short files can't over-pan. `panning_reaches_the_end_of_very_long_lines` drives the real key handler across a 1500-column line and asserts hscroll reaches 1499.

File: src/ui.rs [hunk 4]
```diff
@@ -712,6 +722,19 @@ impl<'a> App<'a> {
                 self.input = Some(InputMode::Summary { buf: String::new() });
                 return None;
             }
+            KeyCode::Char('b') => {
+                self.show_navigator = !self.show_navigator;
+                if !self.show_navigator && self.focus == Focus::Navigator {
+                    self.focus = Focus::Diff;
+                }
+                return None;
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:805]: Fixed in 64171a9: the 'b' arm now calls ensure_cursor_visible before returning, and Event::Resize gets the same reflow (same bug class — a shrinking viewport stranded the cursor identically). `navigator_toggle_reflows_the_viewport` reproduces the stacked-layout case through handle_key and fails without the fix.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 5]
```diff
@@ -1271,6 +1374,99 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+            // Mark the left clip on the first visible cell, preserving its
+            // display width (a wide char's cell becomes "‹ ").
+            for span in line.spans.iter_mut().skip(pinned) {
+                if let Some(first) = span.content.chars().next() {
+                    let rest: String = span.content.chars().skip(1).collect();
+                    let marked = if char_cols(first) >= 2 {
+                        format!("\u{2039} {rest}")
+                    } else {
+                        format!("\u{2039}{rest}")
+                    };
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 703cea6: pan_cap now stops 3 columns short of the widest row — 1 for the ‹ marker plus 2 for a possibly double-width final glyph — so the tail stays inspectable even though the 8-col step can skip intermediate offsets. The long-line test now asserts the final characters render past the marker at max pan. (Note: this commit also carries the fix for the sibling focus finding — both were staged together by mistake; contents are as described in each thread.)

File: src/ui.rs [hunk 6]
```diff
@@ -972,6 +1049,13 @@ impl<'a> App<'a> {
...
                     KeyCode::Char('g') => self.diff.top(),
                     KeyCode::Char('G') => self.diff.bottom(row_count),
+                    KeyCode::Right | KeyCode::Char('L') => {
+                        self.diff.hscroll = (self.diff.hscroll + HSCROLL_STEP).min(self.pan_cap())
+                    }
+                    KeyCode::Left | KeyCode::Char('H') => {
+                        self.diff.hscroll = self.diff.hscroll.saturating_sub(HSCROLL_STEP)
+                    }
+                    KeyCode::Char('0') => self.diff.hscroll = 0,
                     KeyCode::Char('h') | KeyCode::Tab => self.focus = Focus::Navigator,
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 703cea6: h/Tab from the diff now re-shows a collapsed navigator before focusing it, so focus can never land on an invisible pane. Regression test drives Tab through handle_key with the navigator hidden and asserts it is revealed + focused. (Same shared commit as the sibling finding — see its note.)

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 7]
```diff
@@ -1271,6 +1383,99 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+        if dropped {
+            // Mark the left clip on the first visible cell, preserving its
+            // display width (a wide char's cell becomes "‹ ").
+            for span in line.spans.iter_mut().skip(pinned) {
+                if let Some(first) = span.content.chars().next() {
+                    let rest: String = span.content.chars().skip(1).collect();
+                    let marked = if char_cols(first) >= 2 {
+                        format!("\u{2039} {rest}")
+                    } else {
+                        format!("\u{2039}{rest}")
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 719c6bf: zero-width chars following a dropped base char are dropped with it (no orphaned marks attaching to the marker), and ‹ now replaces the first rendered CELL — base char plus trailing combining marks. Tests: a\u{301}b pan-1 yields exactly "‹" at 1 column; a marked surviving cell (e\u{301}) is replaced whole.

File: src/herdr.rs [hunk 8]
```diff
@@ -17,6 +17,20 @@ fn herdr_bin() -> String {
...
+/// Toggle zoom on the pane this process runs in (the review pane calls this
+/// for its `z` key). Best-effort: a failure is the caller's to ignore — zoom
+/// is a convenience, never a correctness matter.
+pub fn zoom_toggle_current() -> Result<()> {
+    let output = Command::new(herdr_bin())
+        .args(["pane", "zoom", "--current", "--toggle"])
+        .output()
+        .context("spawning herdr CLI for zoom")?;
+    if !output.status.success() {
+        bail!("`herdr pane zoom` failed: {}", String::from_utf8_lossy(&output.stderr).trim());
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/herdr.rs:29]: Added in 60083c5: a fake HERDR_BIN_PATH script records argv — asserts the exact 'pane zoom --current --toggle' contract on success and that a nonzero exit surfaces as Err. Env mutation is serialized behind a mutex so parallel tests can't cross-contaminate.

File: src/ui.rs [hunk 9]
```diff
@@ -1069,7 +1162,7 @@ fn draw_footer(frame: &mut Frame, area: Rect, app: &App) {
                 // "where am I" survives and only the key hints get cut.
                 let pos = app.cursor_position().unwrap_or_default();
                 format!(
-                    " {pos} \u{b7} j/k move \u{b7} d/u half page \u{b7} n/p hunk \u{b7} g/G top/bottom \u{b7} h/tab navigator \u{b7} v select \u{b7} c comment \u{b7} x delete \u{b7} {GLOBAL_HINTS}"
+                    " {pos} \u{b7} j/k move \u{b7} \u{2190}/\u{2192} pan \u{b7} d/u half page \u{b7} n/p hunk \u{b7} b files \u{b7} z zoom \u{b7} v select \u{b7} c comment \u{b7} x delete \u{b7} h/tab navigator \u{b7} {GLOBAL_HINTS}"
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:1247]: Fixed in 8a2453b: README now carries the full keyboard+mouse controls table (per CONTRIBUTING's own rule), and the navigator footer advertises 'b hide · z zoom' so both layout controls are discoverable from the initial view.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 10]
```diff
@@ -1271,6 +1383,114 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+        let mut crossing = true; // still owed: orphan marks of the last dropped cell
+        for span in line.spans.iter_mut().skip(pinned) {
+            if col >= hscroll && !pad_next && !crossing {
+                break;
+            }
+            let mut kept = String::new();
+            for c in span.content.chars() {
+                let w = char_cols(c);
+                if col < hscroll {
+                    col += w;
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Partially applied in 4855163. Fixed the real defect: cluster ATOMICITY — a pan boundary inside a ZWJ sequence now drops the whole cluster (ZWJ chains, joined pictographs, skin-tone modifiers, combining marks) with column padding, and a right-clip trims any dangling joiner, so no partial family emoji or marker-fused joiner can render. Tests cover mid-cluster pan and post-ZWJ clip.

Deliberately NOT adopted: counting the sequence as one rendered cell. ratatui measures these spans by per-scalar unicode-width sums (family emoji = 8) when it pads and truncates — matching the render layer's arithmetic is what keeps rows aligned, and terminals themselves disagree on ZWJ cell counts. Diverging from ratatui here would misalign every row containing such a sequence.

**Comment by JonasBaeumer** [comment]
@codex review

File: Cargo.toml [hunk 11]
```diff
@@ -18,6 +18,8 @@ crossterm = "0.28"
 ratatui = "0.29"
 # fancy-regex backend: pure Rust, no oniguruma C dependency
 syntect = { version = "5", default-features = false, features = ["default-fancy"] }
+# terminal display columns (CJK/emoji are 2 cols); pure Rust
+unicode-width = "0.2"
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply Cargo.toml:24]: Justification added to the PR description (Dependency note section): unicode-width 0.2 is pure Rust, already in the tree via ratatui (no new transitive code), and is the same crate ratatui uses — chosen precisely so pan/clip arithmetic matches the render layer. ratatui's own width handling applies to whole spans at render time and is not exposed for the per-scalar boundary decisions panning needs.

File: src/ui.rs [hunk 12]
```diff
@@ -1271,6 +1383,152 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+            if used + cols <= keep {
+                used += cols;
+                continue;
+            }
+            let mut kept = String::new();
+            for c in span.content.chars() {
+                let w = char_cols(c);
+                if used + w > keep {
+                    break;
+                }
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Triage: valid — the right-clip can retain a partial pictograph prefix (family → couple), and the current cleanup only trims trailing joiners/marks, not retained pictographs back to the cluster start. Not fixed in this loop: the 5-round review cap is reached; leaving this thread open for the maintainer. Proposed fix for the next pass: when the clip cut lands inside a cluster (walk back through ZWJ/modifier/mark chains to the cluster head), drop the whole cluster and let the … marker carry the remaining columns.

[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in 23763bacc1e087b5bf65ae8998e1783a496d1959 (`fix: right-clip drops a cut emoji cluster whole`).

Reproduced first: with the old code, a right-clip landing after the family emoji's 2nd pictograph (👩‍👩‍👧‍👦 → cut before the girl) left `"👩\u{200d}👩"` behind after the old trailing-joiner-only cleanup — a couple, exactly as reported.

Fix: `pan_and_clip`'s right-clip now detects a cut mid-cluster (the scalar that missed the budget is a ZWJ/modifier/zero-width mark, or the retained tail itself ends in a ZWJ) and walks the retained tail back through pictographs joined by ZWJs / skin-tone modifiers / zero-width marks to the cluster head, dropping the whole cluster instead of a truncated prefix. Per-scalar width math (matching ratatui) is untouched — only the atomicity of the drop.

New regression test `pan_and_clip_counts_display_columns_not_chars` (extended, in `src/ui.rs`) asserts: right-clip cutting the family mid-cluster leaves zero pictographs and no joiner in the visible content, and total display columns stay ≤ the budget. Verified the test fails without the fix (panics with the couple-emoji content above) and passes with it. All 46 tests green, `cargo build` clean.

[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in 23763ba: a right clip landing mid-cluster now walks the retained tail back to the cluster head (pictographs joined by ZWJs, skin-tone modifiers, zero-width marks) and drops the whole cluster — a family emoji cut after its second member disappears atomically instead of rendering as a couple. Regression assertions added to the pan/clip test (cut before the 3rd member: no pictograph, no joiner survives, width budget held).

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 13]
```diff
@@ -743,15 +803,25 @@ impl<'a> App<'a> {
...
     }
 
+    /// Largest useful horizontal pan for the selected file: three columns
+    /// short of its widest row's pannable width, leaving room for the `‹`
+    /// clip marker (1 col) plus the final glyph even when it is double-width
+    /// (2 cols) — otherwise, at maximum pan the marker replaces the last
+    /// visible character and the 8-col step granularity can make the
+    /// intermediate offsets that would reveal it unreachable.
+    fn pan_cap(&self) -> usize {
+        self.pan_limit.get(&self.nav.selected).copied().unwrap_or(0).saturating_sub(3)
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in 589a3c0fb91d21ace67a9932a664e4c5335f5371 (`fix: derive pan_cap's reserve from the actual trailing glyph width`).

Reproduced first: a 5-char ASCII line ("abcde"), real gutter+marker pinned width 11, in a 13-column pane (2 content columns after the gutter, matching your report) — with the old flat -3 reservation, pan_cap capped at 2, and the right-clip step further reduced the panned tail to `"‹…"`, so 'e' was never visible.

Fix: `pan_cap_for_rows` now reserves `1 (marker) + trailing_cell_width` — the actual display width (1 or 2) of the widest row's own last nonzero-width scalar — instead of a flat 3 that assumed a double-width glyph. For this ASCII case that tightens the cap by exactly the 1 column needed: at the new cap the tail becomes `"‹e"`, fully within the 13-column budget, no right-clip needed.

New test `narrow_panes_can_still_reach_the_final_ascii_character` reproduces your exact scenario and fails pre-fix with `"‹…"`. The existing 1500-char regression test's hardcoded expectations (1497 / "xx") encoded the old over-reservation and are updated to the corrected values (1498 / a single trailing "x") — its actual invariant (marker lands before the final glyph, not on it) is unchanged and still holds. All 47 tests green, `cargo build` clean.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 14]
```diff
@@ -972,7 +1088,20 @@ impl<'a> App<'a> {
                     }
                     KeyCode::Char('g') => self.diff.top(),
                     KeyCode::Char('G') => self.diff.bottom(row_count),
-                    KeyCode::Char('h') | KeyCode::Tab => self.focus = Focus::Navigator,
+                    KeyCode::Right | KeyCode::Char('L') => {
+                        self.diff.hscroll = (self.diff.hscroll + HSCROLL_STEP).min(self.pan_cap())
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in 89f145bb229c216b6f8e23a3fea763820c4d1495 (`fix: keep regional-indicator flag pairs atomic in pan and right-clip` — this commit bundles two independent fixes since both were pushed together; apologies for not splitting them).

Reproduced first: a file with a 6-char row ("abcdef") and a 200-char row in the same file (so the long row sets a big `pan_cap`), rendered in a 14-column pane (11 pinned gutter/marker columns, 2 content columns — matching your report). With the old fixed 8-column step, the first Right press jumped straight from hscroll=0 (showing "ab…") to hscroll=8, past the short row's entire 6-column width, landing on "fully panned off, empty" — 'c','d','e','f' were never reachable at any hscroll value.

Fix: added `next_pan_stop`, used by both the keyboard Right/L handler and the mouse ScrollRight handler. It takes the full `HSCROLL_STEP` jump UNLESS some row's pannable width sits strictly inside that jump, in which case it steps by a single column instead — so a short row's remaining content is revealed one column at a time until the jump would no longer skip over anything, at which point it resumes full-speed 8-column jumps (long lines with no such conflict are unaffected).

New test `short_rows_stay_reachable_when_a_longer_row_sets_a_big_pan_cap` reproduces your exact scenario, asserting 'c','d','e','f' all surface across the press sequence (fails pre-fix — the set stays empty), and that Right still reaches the file-wide `pan_cap` in a bounded number of presses afterward (no navigation slowdown on long lines). All 49 tests green, `cargo build` clean.

File: src/ui.rs [hunk 15]
```diff
@@ -1271,6 +1419,188 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+            for c in span.content.chars() {
+                let w = char_cols(c);
+                if col < hscroll {
+                    col += w;
+                    dropped = true;
+                    if col > hscroll {
+                        pad_cols += col - hscroll; // wide char straddled
+                    }
+                    prev_zwj = c == '\u{200d}';
+                } else if crossing && (w == 0 || prev_zwj || is_emoji_modifier(c)) {
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in 89f145bb229c216b6f8e23a3fea763820c4d1495 (`fix: keep regional-indicator flag pairs atomic in pan and right-clip`).

You're right that the cluster-tail predicate had no concept of regional-indicator (RI) pairing — verified directly: RIGHT-CLIP cutting a flag mid-pair (first RI kept, second dropped by the budget) left the first RI standing alone in the output (reproduced `got "🇺"` before the fix). Fixed by extending the walk-back trigger to detect a cut landing on the second scalar of an odd (unpaired) trailing RI run — the existing walk-back loop needed no changes, it already pops exactly one trailing non-glue char once primed.

One nuance on the PAN-left side specifically: a bare flag at the very start of panned content (your literal example) turns out to be masked by `pan_and_clip`'s own marker-replacement step — it always overwrites whatever the first surviving character is with `‹`, and for a lone 2-scalar RI pair that first surviving character is always the orphan itself, so the unfixed and fixed output are byte-for-byte identical in that specific shape (verified — disabling just the pan-left half of the fix did not make my initial test fail). The underlying predicate gap is still real, though: I confirmed it's independently observable when the flag is preceded by other crossing-eligible content in the same run (e.g. a CJK char + ZWJ immediately before the flag, so pad is already owed when the flag is reached) — there the orphan survives visibly past the marker. Fixed the predicate on both pan and right-clip since they're the same gap, and the regression test covers both: the right-clip case directly, and the pan case via the CJK+ZWJ-bridged construction that actually discriminates.

All 49 tests green, `cargo build` clean.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 16]
```diff
@@ -743,15 +843,48 @@ impl<'a> App<'a> {
...
+    /// short-row conflict still jump the full step.
+    fn next_pan_stop(&self, current: usize) -> usize {
+        let target = current + HSCROLL_STEP;
+        let overshoots_a_row = self
+            .row_cache
+            .get(&self.nav.selected)
+            .into_iter()
+            .flatten()
+            .map(pannable_cols)
+            .any(|cols| cols > current && cols < target);
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in 220e4988af57f76b077fd2274e12c5717b190960 (`fix: treat an exact-step row boundary as an overshoot too`).

You're right — the strict `cols < target` in `next_pan_stop` excluded exactly the row it needed to catch when a row's pannable width equals `current + HSCROLL_STEP` precisely. Reproduced: an 8-char row + a long row (forcing a big `pan_cap`) in a 14-column pane — the first Right press jumped straight from hscroll=0 to hscroll=8 (that row already fully panned off), skipping every offset that would have revealed 'c' through 'h'.

Fix: changed the comparison to `cols <= target`, so a row landing exactly on the boundary triggers the same single-column fine-stepping as one strictly inside it.

New test `short_rows_stay_reachable_at_an_exact_step_boundary` reproduces this (verified it fails pre-fix — the visible-character set stays empty across 8 presses; note the test uses an empty hunk header, since even a short nonempty one would independently trigger fine-stepping and mask the specific boundary being tested). All 51 tests green, `cargo build` clean.

File: src/ui.rs [hunk 17]
```diff
@@ -549,15 +564,81 @@ struct App<'a> {
...
+/// nothing to protect, so no need to reserve for a wide glyph that isn't
+/// there).
+fn trailing_cell_width(line: &Line) -> usize {
+    line.spans
+        .iter()
+        .skip(pinned_spans(line))
+        .flat_map(|s| s.content.chars())
+        .filter(|c| char_cols(*c) > 0)
+        .last()
+        .map(char_cols)
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in c7fd556a2efee4cdd05014c7f34e9ccd4bd26d4f (`fix: reserve the whole trailing regional-indicator flag pair`).

Verified the mechanism precisely: for "abcdef🇺🇸" (pannable_cols=8), the old `trailing_cell_width` returned 1 (just the last RI scalar), giving `pan_cap=6`. At hscroll=6 the pan boundary lands cleanly BEFORE the flag (not mid-pair, so the pairing-atomicity fix doesn't even trigger) — both RIs get kept, but the marker-replacement step then overwrites the FIRST RI with `‹` (it only ever protects trailing zero-width marks on the character it replaces, not a second cluster scalar), leaving the second RI orphaned: `"‹🇸"`.

Fix: `trailing_cell_width` now checks whether the trailing RI completes an even-length run in the pannable content (a full pair, not an odd one out) and reserves 2 cols for it in that case, giving `pan_cap=5` here — the marker then lands on the unrelated 'f' instead, and the whole flag survives together: `"‹🇺🇸"`.

New test `pan_cap_protects_a_trailing_flag_pair_whole` asserts both the corrected reserve (`pan_cap()==5`) and that both RI scalars appear together at max pan; verified it fails pre-fix (computed cap 6, and disabling just the parity check reproduces the orphaned-second-RI output). All 51 tests green, `cargo build` clean.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 18]
```diff
@@ -1271,6 +1472,222 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+        }
+        if dropped {
+            // Mark the left clip on the first visible CELL — the first
+            // nonzero-width char plus its trailing combining marks —
+            // preserving its display width (a wide cell becomes "‹ ").
+            for span in line.spans.iter_mut().skip(pinned) {
+                if let Some(first) = span.content.chars().next() {
+                    let cell_width = char_cols(first).max(1);
+                    let rest: String = span
+                        .content
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Confirmed and fixed in f36cca9893b2ced241cb6a8c054e49a60c831459 (`fix: replace the whole leading cluster when marking a pan cut`).

Reproduced your exact example: panning "abcdefgh🇺🇸xyz" by 8 landed the boundary right at the flag (a clean cut, not mid-pair — the pairing-atomicity fix from a couple rounds ago doesn't even trigger here), then the marker-replacement step swapped only the first regional indicator for `‹`, producing `"‹🇸xyz"` — exactly as reported.

Fix: added `leading_cluster_len`, which scans FORWARD from the first surviving scalar using the same cluster rules already enforced elsewhere in this file (zero-width marks, skin-tone modifiers, ZWJ continuations, RI pairs), and the marking step now replaces the WHOLE leading cluster with `‹` + padding spaces sized to the cluster's total display width — the same convention already used for a single wide char, just generalized.

New test `pan_marker_replaces_the_whole_leading_cluster_not_just_its_first_scalar` covers both your RI-flag example (verified it fails pre-fix, reproducing `"‹🇸xyz"` exactly) and an analogous leading ZWJ family-emoji case. All 52 tests green — the other 51 are byte-identical to before (the new logic reduces to the same single-scalar-plus-marks behavior for every case that isn't a multi-scalar cluster), `cargo build` clean.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 19]
```diff
@@ -1271,6 +1472,253 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+        return 0;
+    };
+    if is_regional_indicator(first) && chars.get(1).is_some_and(|&c| is_regional_indicator(c)) {
+        return 2;
+    }
+    let mut i = 1;
+    let mut prev_zwj = first == '\u{200d}';
+    while let Some(&c) = chars.get(i) {
+        if char_cols(c) == 0 || is_emoji_modifier(c) || prev_zwj {
+            prev_zwj = c == '\u{200d}';
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Fixed in aed8ca8 — and structurally, per your framing: pan_and_clip now segments by UAX #29 extended grapheme clusters (unicode-segmentation 1.13, which implements GB9c so KA+virama+SSA is one cluster). The whole hand-rolled series — ZWJ chains, flag pairing, walk-backs, joiner cleanup — is deleted in favor of 'a cluster drops whole at any boundary'. Widths are unicode-width 0.2 string widths per grapheme, the exact pair ratatui 0.29 pins, so column math matches rendering verbatim. Regression test covers the conjunct; family/flag tests restated as whole-or-gone invariants. Dependency note added to the PR description.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 20]
```diff
@@ -743,15 +837,54 @@ impl<'a> App<'a> {
...
+        // revealed that row's middle/tail (current+1..target-1) was never
+        // visited, and `target` itself is already "fully panned off, empty"
+        // for that row.
+        let overshoots_a_row = self
+            .row_cache
+            .get(&self.nav.selected)
+            .into_iter()
+            .flatten()
+            .map(pannable_cols)
+            .any(|cols| cols > current && cols <= target);
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:886]: Fixed in 7cbfad1: the pan step now caps at half the visible code columns (minimum 1), so a 2-column pane advances column by column and every offset — including a short row's middle — is reachable. Normal panes keep the 8-column stride plus the existing short-row fine-step rule. Regression test drives Right through handle_key at a 16-column terminal and asserts single-column advances.

File: src/ui.rs [hunk 21]
```diff
@@ -549,15 +565,74 @@ struct App<'a> {
...
+/// reach (Codex P1: a flat -3 reservation cost narrow panes exactly the
+/// one column a single-width trailing glyph didn't need reserved).
+/// When several rows tie for the widest, reserves for whichever of THEM
+/// has the widest trailing glyph, so panning to this cap is safe for all.
+fn pan_cap_for_rows(rows: &[Line]) -> usize {
+    let max_cols = rows.iter().map(pannable_cols).max().unwrap_or(0);
+    let reserve = rows
+        .iter()
+        .filter(|r| pannable_cols(r) == max_cols)
+        .map(|r| 1 + trailing_cell_width(r))
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:624]: Rebutted as unsatisfiable at the stated width: with two code columns, the ‹ marker (1 col) plus a double-width glyph (2 cols) need three columns — no cap value can render both, and dropping the marker would silently hide that content exists to the left. Reaching such a pane requires a terminal ~14 columns wide (the pinned gutter alone is 12): the side-by-side layout can't get there (navigator minimum keeps the diff ≥ ~38 cols) and the stacked layout needs the whole terminal at 14 cols. With 7cbfad1's single-column stepping, a reviewer in that extreme can pan to see the glyph's left half and the … marker; the complete glyph is simply wider than the available viewport, same as any content wider than its window. Not special-casing this; the thread stays open for the maintainer's judgment.
